small_language_model_pretraining/nemo/README.md

```diff
@@ -0,0 +1,220 @@
+
+```bash
+export C4_PATH=""
+export MERGED_C4_PATH=""
+# more information about this knob can be found in consolidate_data.sh
+export N_VALIDATION_SAMPLES=91205
+
+bash utils/consolidate_data.sh
+```
+
+
+### Tokenizer
+
+We are using the Llama 3.1 8B tokenizer. You can run `utils/download_hf_llama3.sh` to download it. 
+
+
+After the data consolidation is done, we can run this [script](./utils/preprocess.sh) to perform preprocessing. To run the preprocessing script, we need to use the following commands: 
+
+```bash
+# fill in the built container path here
+export CONT_IMAGE_URL=""
+# pass in the folder path that contains the Llama tokenizer here
+# please refer to the tokenizer section above for more details
+export TOKENIZER_PATH=""
+# pass in the merged file path here
+export MERGED_C4_PATH=""
+# this path is used for storing the preprocessed .bin and .idx files
+export PREPROCESSED_PATH=""
+
+# Extra Slurm-related arguments can be provided here
+sbatch preprocess.sh
+```
+
+Warning! If you receive an error message of file not found, look into where `preprocess_data_for_megatron.py` is located in your path. 
+
+
+#### Training and test data separation
+
+We use the default split from the C4 dataset. This means that we use `c4-train.<x>-of-01024.json.gz` files (where `768 <= x <= 1023`) for training, and we use our customized `c4-validation-91205-samples.en.json.gz`, which contains the first 91205 samples from the unshuffled C4 validation dataset, for evaluation. 
+
+Notice here that we are using the first 5760 sequences (47,185,920 tokens) from the validation dataset to perform the validation. According to our experiments, the first 91205 samples from the unshuffled C4 dataset yields 47,186,855 tokens, which is the smallest amount of samples needed to yield 47,185,920 tokens. Thus, we have chosen the first 91205 samples as our validation dataset. 
+
+#### Training data order
+
+We randomly shuffle the **last 256 of 1024 shards** for the benchmarking area.
+
+#### Test data order
+
+We use the first 5,760 sequences (91,205 untokenized samples) in the validation dataset for validation. We **do not shuffle** the validation dataset. 
+
+# 4. Model
+### Publication/Attribution
+
+The model largely follows the Llama 3.1 8B [paper](https://arxiv.org/abs/2407.21783). 
+
+### Model details
+
+| Config | Value |
+| :-- | :-- | 
+| Embedding | RoPE + parameter adjustments |
+| # Layers | 32 | 
+| Attention Type | GQA |
+| # Attn Heads | 32 | 
+| Key/Value Heads | 8 | 
+| Model Dimension | 4096 |
+| FFN Dimension | 14336 |
+| Activation | SwiGLU | 
+| Normalization | RMSNorm |  
+| Tokenizer | Llama tokenizer |
+| Vocab size | 128,000 |  
+| Context Length | 8192 |
+
+
+### Checkpoint download
```

Comment:
Remove this section and mention that the benchmark trains from random weights so checkpoint download is not needed